Framework/API/src/IndexProperty.cpp

```diff
@@ -0,0 +1,147 @@
+#include "MantidAPI/IndexProperty.h"
+#include "MantidAPI/MatrixWorkspace.h"
+#include "MantidIndexing/GlobalSpectrumIndex.h"
+#include "MantidIndexing/IndexInfo.h"
+#include "MantidIndexing/SpectrumNumber.h"
+
+namespace Mantid {
+namespace API {
+IndexProperty::IndexProperty(const std::string &name,
+                             const IWorkspaceProperty &workspaceProp,
+                             const IndexTypeProperty &indexTypeProp,
+                             Kernel::IValidator_sptr validator)
+    : ArrayProperty(name, validator, Kernel::Direction::Input),
+      m_workspaceProp(workspaceProp), m_indexTypeProp(indexTypeProp), m_min(0),
+      m_max(0), m_indices(0), m_indicesExtracted(false) {}
+
+std::string IndexProperty::isValid() const {
+  std::string error;
+
+  try {
+    getIndices();
+  } catch (std::runtime_error &e) {
+    error = e.what();
+  } catch (std::out_of_range &) {
+    error = "Indices provided to IndexProperty are out of range.";
+  } catch (std::logic_error &) {
+    error = "Duplicate indices supplied to IndexProperty.";
+  }
```

Comment:
Probably this is sufficient for the first iteration but I foresee complaints about not providing the offending indices (this would need to change the error handling in `IndexInfo`/`SpectrumNumberTranslator`.